server/vmap/trillian_map_server.go

```diff
@@ -74,24 +74,33 @@ func (t *TrillianMapServer) GetLeaves(ctx context.Context, req *trillian.GetMapL
 
 	smtReader := merkle.NewSparseMerkleTreeReader(req.Revision, hasher, tx)
 
-	leaves, err := tx.Get(ctx, req.Revision, req.Index)
-	if err != nil {
-		return nil, err
-	}
-	glog.Infof("%v: wanted %v leaves, found %v", mapID, len(req.Index), len(leaves))
+	// TODO: we need to return one response per request.
+	// Get only returns items for things that exist.
 
-	resp := &trillian.GetMapLeavesResponse{
-		MapLeafInclusion: make([]*trillian.MapLeafInclusion, len(leaves)),
-	}
-	for i, leaf := range leaves {
-		proof, err := smtReader.InclusionProof(ctx, req.Revision, leaf.Index)
+	inclusions := make([]*trillian.MapLeafInclusion, 0, len(req.Index))
+	for i, index := range req.Index {
+		proof, err := smtReader.InclusionProof(ctx, req.Revision, index)
 		if err != nil {
 			return nil, err
 		}
-		// Copy the leaf from the iterator, which gets overwritten
-		value := leaf
+
+		// Create placeholder leaf for non-existant leaves.
+		leaf := &trillian.MerkleLeaf{
+			Index: index,
+		}
+
+		// Fetch the leaf if it exists.
+		leafs, err := tx.Get(ctx, req.Revision, [][]byte{index})
+		if err != nil {
+			return nil, err
+		}
+		for l := range leafs {
+			leaf = l
+			break
+		}
+
```

Comment:
This might deserve a GVC discussion, but because this is a crypto API for an untrusted server, the behavior might need to diverge slightly from what a typical API might do. 

- Leaf not in storage needs to return a proof of non-inclusion, unlike typical APIs. Suppose the server has updated a leaf and wants to hide it from a particular user. The server could return `NotFound` instead of the real value. To prevent this, the server must return proofs, even for `NotFound`. 
- `len(req.Index) == 0` is a syntactically valid request. Treating len == 0 as a special case means clients need to special case their logic as well, and special cases are ugly. 
- `len(index) == 0 != tree.IndexSize` is certainly an error case.